test/core/stream/batch_t.cpp

```diff
@@ -0,0 +1,192 @@
+BOOST_AUTO_TEST_CASE(test_batch_take)
+{
+    using hipipe::stream::batch_t;
+
+    batch_t batch1;
+    batch1.insert_or_assign<Int>(std::vector<int>{0, 1, 2});
+    batch1.insert_or_assign<Double>(std::vector<double>{0., 1., 2.});
+    batch_t batch2 = batch1.take(2);
+    BOOST_TEST(batch1.extract<Int>() == std::vector<int>({2}));
+    BOOST_TEST(batch1.extract<Double>() == std::vector<double>({2.}));
+    BOOST_TEST(batch2.extract<Int>() == std::vector<int>({0, 1}));
+    BOOST_TEST(batch2.extract<Double>() == std::vector<double>({0., 1.}));
+    batch_t batch3 = batch1.take(1);
+    BOOST_TEST(batch1.extract<Int>() == std::vector<int>({}));
+    BOOST_TEST(batch1.extract<Double>() == std::vector<double>({}));
+    BOOST_TEST(batch3.extract<Int>() == std::vector<int>({2}));
+    BOOST_TEST(batch3.extract<Double>() == std::vector<double>({2.}));
+}
+
+
+BOOST_AUTO_TEST_CASE(test_batch_take_move_only)
+{
+    using hipipe::stream::batch_t;
+
+    batch_t batch1;
+    batch1.insert_or_assign<Unique>();
+    batch1.extract<Unique>().push_back(std::make_unique<int>(1));
+    batch1.extract<Unique>().push_back(std::make_unique<int>(2));
+    batch_t batch2 = batch1.take(1);
+    BOOST_TEST(batch1.batch_size() == 1);
+    BOOST_TEST(batch2.batch_size() == 1);
+    BOOST_TEST(*batch1.extract<Unique>().at(0) == 2);
+    BOOST_TEST(*batch2.extract<Unique>().at(0) == 1);
+}
+
+
+BOOST_AUTO_TEST_CASE(test_batch_take_throws_error)
+{
+    using hipipe::stream::batch_t;
+
+    batch_t batch1;
+    batch1.insert_or_assign<Int>(std::vector<int>{0, 1, 2});
+    batch1.insert_or_assign<Double>(std::vector<double>{0., 1., 2.});
+    BOOST_CHECK_THROW(batch1.take(4), std::runtime_error);
+}
+
+
+BOOST_AUTO_TEST_CASE(test_batch_push_back)
+{
+    using hipipe::stream::batch_t;
+
+    batch_t batch1;
+    batch1.insert_or_assign<Int>(std::vector<int>{0, 1, 2});
+    batch_t batch2;
+    batch2.insert_or_assign<Int>(std::vector<int>{3, 4});
+    batch2.insert_or_assign<Double>(std::vector<double>{0., 1., 2.});
+    batch_t batch3;
+    batch1.push_back(std::move(batch2));
+    BOOST_TEST(batch1.extract<Int>() == std::vector<int>({0, 1, 2, 3, 4}));
+    BOOST_TEST(batch1.extract<Double>() == std::vector<double>({0., 1., 2.}));
+    batch1.push_back(std::move(batch3));
+    BOOST_TEST(batch1.extract<Int>() == std::vector<int>({0, 1, 2, 3, 4}));
+    BOOST_TEST(batch1.extract<Double>() == std::vector<double>({0., 1., 2.}));
+}
+
+
+BOOST_AUTO_TEST_CASE(test_batch_push_back_move_only)
+{
+    using hipipe::stream::batch_t;
+
+    batch_t batch1;
+    batch1.insert_or_assign<Unique>();
+    batch1.extract<Unique>().push_back(std::make_unique<int>(1));
+    batch1.extract<Unique>().push_back(std::make_unique<int>(2));
+    batch_t batch2;
+    batch2.insert_or_assign<Unique>();
+    batch2.extract<Unique>().push_back(std::make_unique<int>(3));
+    batch1.push_back(std::move(batch2));
+    BOOST_TEST(batch1.batch_size() == 3);
+    BOOST_TEST(batch2.batch_size() == 0);
```

Comment:
Please remove this line. The `batch2` is already "moved" and the documentation stays that using it further causes undefined behavior.